server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala

```diff
@@ -0,0 +1,136 @@
+  private val publisher: Promise[Publisher[NettyWebSocketFrame]] = Promise[Publisher[NettyWebSocketFrame]]()
+  @volatile private var subscription: Subscription = _
+
+  private val logger = LoggerFactory.getLogger(getClass.getName)
+
+  override def onSubscribe(s: Subscription): Unit = {
+    // Not really that unsafe. Subscriber creation doesn't do any IO, only initializes an AtomicReference in an initial state.
+    subscriber = dispatcher.unsafeRunSync(
+      // If bufferSize > 1, the stream may stale and not emit responses until enough requests are buffered
+      StreamSubscriber[F, NettyWebSocketFrame](bufferSize = 1)
+    )
+    subscription = new NonCancelingSubscription(s)
+    val in: Stream[F, NettyWebSocketFrame] = subscriber.sub.stream(Applicative[F].unit)
+    val sttpFrames = in.map { f =>
+      val sttpFrame = nettyFrameToFrame(f)
+      f.release()
+      sttpFrame
+    }
+    val stream: Stream[F, NettyWebSocketFrame] =
+      optionallyConcatenateFrames(sttpFrames, o.concatenateFragmentedFrames)
+        .map(f =>
+          o.requests.decode(f) match {
+            case x: DecodeResult.Value[REQ]    => x.v
+            case failure: DecodeResult.Failure => throw new WebSocketFrameDecodeFailure(f, failure)
+          }
+        )
+        .through(pipe)
+        .map(r => frameToNettyFrame(o.responses.encode(r)))
+        .onFinalizeCaseWeak {
+          case ExitCase.Succeeded =>
+            Sync[F].delay { val _ = wsCompletedPromise.setSuccess() }
+          case ExitCase.Errored(t) =>
+            Sync[F].delay(wsCompletedPromise.setFailure(t)) >> Sync[F].delay(logger.error("Error occured in WebSocket channel", t))
+          case ExitCase.Canceled =>
+            Sync[F].delay { val _ = wsCompletedPromise.cancel(true) }
+        }
+        .append(fs2.Stream(frameToNettyFrame(WebSocketFrame.close)))
+
+    // Trigger listening for WS frames in the underlying fs2 StreamSubscribber
+    subscriber.sub.onSubscribe(subscription)
+    // Signal that a Publisher is ready to send result frames
+    publisher.success(StreamUnicastPublisher(stream, dispatcher))
+  }
+
+  override def onNext(t: NettyWebSocketFrame): Unit = {
+    subscriber.sub.onNext(t)
+  }
+
+  override def onError(t: Throwable): Unit = {
+    subscriber.sub.onError(t)
+    if (!wsCompletedPromise.isDone()) {
+      val _ = wsCompletedPromise.setFailure(t)
+    }
+  }
+
+  override def onComplete(): Unit = {
+    subscriber.sub.onComplete()
+    if (!wsCompletedPromise.isDone()) {
+      val _ = wsCompletedPromise.setSuccess()
+    }
+  }
+
+  override def subscribe(s: Subscriber[_ >: NettyWebSocketFrame]): Unit = {
+    // A subscriber may come to read from our internal Publisher. It has to wait for the Publisher to be initialized.
+    publisher.future.onComplete {
+      case Success(p) =>
+        p.subscribe(s)
+      case Failure(ex) =>
+        subscriber.sub.onError(ex)
+        subscription.cancel
```

Comment:
Cosmetic: use `cancel()` (with parens) as it's a side-effectful invocation.

Reply:
Actually, the `subscription` is a `NonCancelingSubscription` so this invocation does nothing.

BTW, why is it non-canceling?